evm-precompiles/erc20-peg/src/lib.rs

```
@@ -0,0 +1,124 @@
+			return Err(err);
+		}
+
+		match selector {
+			Action::Withdraw => Self::withdraw(input, gasometer, &context.caller),
+		}
+	}
+}
+
+impl<T> Erc20PegPrecompile<T> {
+	pub fn new() -> Self {
+		Self(PhantomData)
+	}
+}
+
+impl<T> Erc20PegPrecompile<T>
+where
+	T: frame_system::Config
+		+ crml_erc20_peg::Config
+		+ pallet_evm::Config
+		+ Erc20IdConversion<EvmId = Address, RuntimeId = AssetId>,
+{
+	/// erc20-peg withdrawal
+	fn withdraw(input: &mut EvmDataReader, gasometer: &mut Gasometer, caller: &H160) -> EvmResult<PrecompileOutput> {
+		// Parse input.
+		input.expect_arguments(gasometer, 3)?;
+		let fee_asset_id: Address = input.read::<Address>(gasometer)?.into();
+		// the given `input_asset` address is not a valid (derived) generic asset address
+		// it is not supported by cennzx
+		let asset_id = T::evm_id_to_runtime_id(fee_asset_id).ok_or(gasometer.revert("unsupported asset"))?;
+		let withdraw_amount: U256 = input.read::<U256>(gasometer)?.into();
+		let withdraw_amount = scale_wei_to_4dp(withdraw_amount.unique_saturated_into());
+		let beneficiary: H160 = input.read::<Address>(gasometer)?.into();
+
+		gasometer.record_cost(RuntimeHelper::<T>::db_read_gas_cost())?;
+		let caller: T::AccountId = T::AddressMapping::into_account_id(*caller);
+		let event_proof_id = crml_erc20_peg::Module::<T>::do_withdrawal(
+			caller,
+			asset_id,
+			withdraw_amount,
+			beneficiary,
+			WithdrawCallOrigin::Evm,
+		);
+
+		// Build output.
+		match event_proof_id {
+			Ok(proof_id) => Ok(PrecompileOutput {
+				exit_status: ExitSucceed::Returned,
+				cost: gasometer.used_gas(),
+				output: EvmDataWriter::new().write(U256::from(proof_id)).build(),
+				logs: Default::default(),
+			}),
+			Err(_) => Ok(PrecompileOutput {
+				exit_status: ExitSucceed::Returned,
+				cost: gasometer.used_gas(),
+				output: EvmDataWriter::new().build(),
+				logs: Default::default(),
+			}),
+		}
```

don't hide the error state, should be a PrecompileFailure if it failed